### PR TITLE
Issue #16: Add comment support

### LINKED ISSUE #16: Add comment support
Current state of AST:

```bash

➜  checkstyle-antlr4 git:(issue-7) cat /home/nick/Desktop/Tester/src/Test.java
public  class Test {
    /**
     * This is a javadoc comment.
     */
    void doStuff()/*This is a block comment*/ {
        // This is a line comment
    }
}

➜  checkstyle-antlr4 git:(issue-7) java -jar target/checkstyle-8.45-SNAPSHOT-all.jar -T /home/nick/Desktop/Tester/src/Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:8]
|--IDENT -> Test [1:14]
`--OBJBLOCK -> OBJBLOCK [1:19]
    |--LCURLY -> { [1:19]
    |--METHOD_DEF -> METHOD_DEF [5:4]
    |   |--MODIFIERS -> MODIFIERS [5:4]
    |   |--TYPE -> TYPE [5:4]
    |   |   `--LITERAL_VOID -> void [5:4]
    |   |--IDENT -> doStuff [5:9]
    |   |--LPAREN -> ( [5:16]
    |   |--PARAMETERS -> PARAMETERS [5:17]
    |   |--RPAREN -> ) [5:17]
    |   `--SLIST -> { [5:46]
    |       `--RCURLY -> } [7:4]
    `--RCURLY -> } [8:0]

```

What I would expect:

```bash

CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:8]
|--IDENT -> Test [1:14]
`--OBJBLOCK -> OBJBLOCK [1:19]
    |--LCURLY -> { [1:19]
    |--METHOD_DEF -> METHOD_DEF [5:4]
    |   |--MODIFIERS -> MODIFIERS [5:4]
    |   |--TYPE -> TYPE [5:4]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [2:4]
    |   |   |   |--COMMENT_CONTENT -> *\n     * This is a javadoc comment.\n      [2:6]
    |   |   |   `--BLOCK_COMMENT_END -> */ [4:5]
    |   |   `--LITERAL_VOID -> void [5:4]
    |   |--IDENT -> doStuff [5:9]
    |   |--LPAREN -> ( [5:16]
    |   |--PARAMETERS -> PARAMETERS [5:17]
    |   |--RPAREN -> ) [5:17]
    |   |--BLOCK_COMMENT_BEGIN -> /* [5:18]
    |   |   |--COMMENT_CONTENT -> This is a block comment [5:20]
    |   |   `--BLOCK_COMMENT_END -> */ [5:42]
    |   `--SLIST -> { [5:46]
    |       |--SINGLE_LINE_COMMENT -> // [6:8]
    |       |   `--COMMENT_CONTENT ->  This is a line comment\n [6:10]
    |       `--RCURLY -> } [7:4]
    `--RCURLY -> } [8:0]

```

### PR DESCRIPTION
**Pull request by nmancus1**
Closes #16.

Previous commits will be dropped after #7 is merged.

### AST Regression Reports

These were ran normally, **with** comments. Report generation took 35 minutes, same as master branch  (ANTLR2) would.

Only differences between this report and the one at #7:
https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/compilable-test-inputs/index.html#A806
https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/compilable-test-inputs/index.html#A807
https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/compilable-test-inputs/index.html#A808
https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/compilable-test-inputs/index.html#A809
https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/compilable-test-inputs/index.html#A846
https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/compilable-test-inputs/index.html#A847
https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/compilable-test-inputs/index.html#A848
https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/compilable-test-inputs/index.html#A849

All differences are only because the comments are nested inside of structures that are changed in existing PR's in the base repo, so the report is clean.

https://nmancus1.github.io/antlr4-ast-diffs-issue-16/diff-6-29-15_30/index.html

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java [hunk 1]
```diff
@@ -62,11 +63,17 @@
...
+     */
+    private List<Token> hiddenBefore;
 
-    /** Hidden (comment) AST following this DetailAST. */
-    private CommonToken hiddenAfter;
+    /**
+     * All tokens on COMMENTS channel to the right of the current token up to the
+     * next token on the DEFAULT_TOKEN_CHANNEL.
+     */
+    private List<Token> hiddenAfter;
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:76]: Now we can just traverse this list instead of iterating through linked tokens.

File: src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java [hunk 2]
```diff
@@ -26,6 +29,12 @@
  */
 public final class JavaAstVisitor extends JavaParserBaseVisitor<DetailAstImpl> {
 
+    private final BufferedTokenStream tokens;
+
+    public JavaAstVisitor(CommonTokenStream tokenStream) {
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java:34]: New constructor gives us the ability to access the Token stream in the `JavaAstVisitor` class, so that we can add the tokens from the `COMMENTS` channel to the corresponding AST.

File: src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java [hunk 3]
```diff
@@ -262,8 +271,8 @@ public DetailAstImpl visitClassBlock(JavaParser.ClassBlockContext ctx) {
         // If block is static, we create imaginary STATIC_INIT node
         if (ctx.LITERAL_static() != null) {
             final DetailAstImpl temp = classBlock;
-            classBlock = createImaginary(TokenTypes.STATIC_INIT,
-                    ctx.LITERAL_static());
+            classBlock = create(TokenTypes.STATIC_INIT, (Token) ctx.LITERAL_static().getPayload());
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java:274]: Had to change this, we were copying the node AFTER adding the `COMMENTS` tokens, and duplicating comments.

File: src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java [hunk 4]
```diff
@@ -1914,9 +1923,21 @@ private static DetailAstImpl create(int tokenType, Token startToken) {
      * @param token the token to build the DetailAstImpl from
      * @return new DetailAstImpl of given type
      */
-    private static DetailAstImpl create(Token token) {
+    private DetailAstImpl create(Token token) {
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java:1926]: A few methods had to change, since we access the instance of `JavaAstVisitor` when we get the comment tokens.

File: src/main/java/com/puppycrawl/tools/checkstyle/JavaParser.java [hunk 5]
```diff
@@ -135,20 +140,23 @@ public static DetailAST appendHiddenCommentNodes(DetailAST root) {
...
             DetailAST currentSibling = curNode;
-            while (tokenBefore != null) {
-                final DetailAST newCommentNode =
-                         createCommentAstFromToken(tokenBefore);
-
-                ((DetailAstImpl) currentSibling).addPreviousSibling(newCommentNode);
-
-                if (currentSibling == result) {
-                    result = newCommentNode;
+            if (curNode.getHiddenBefore() != null) {
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/JavaParser.java:144]: Really not much changed here, still the same idea when we add nodes.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4 [hunk 6]
```diff
@@ -287,8 +307,39 @@ ELLIPSIS:                '...';
...
+     '*/'
+    {
+        // Trim delimiters from comment text
+        setText(getText().substring(2, getText().length() - 2));
+        // Report comment
+        mCommentListener.reportBlockComment("/*", startLine, startCol,
+            _interp.getLine(), _interp.getCharPositionInLine() - 2);
+        startLine = startCol = -1;
+    } -> channel(COMMENTS);
+
```
Comment (nmancus1) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4:328]: Very little has changed in how we handle this in the Lexer; set the text, report the comment, etc.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4 [hunk 7]
```diff
@@ -287,8 +307,39 @@ ELLIPSIS:                '...';
...
+SINGLE_LINE_COMMENT:
+    // Match single line comment delimiter
+    '//'
+    {
+        // Report comment
+        mCommentListener.reportSingleLineComment("//",
+            _tokenStartLine, _tokenStartCharPositionInLine);
+    }
+     // Match content, including terminating newline (or nothing if EOF)
+    ~[\r\n]* ( '\n' | '\r' ('\n')? | /** nothing */ )
```
Comment (nmancus1) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4:338]: See

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammar/antlr4/InputAntlr4AstRegressionCassandraInputWithComments.java [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+        currentTest.appendChild(nested);
+
+    }
+
+    public void testAssumptionFailure(final Test test, final Throwable failure) {
+        formatSkip(test, failure.getMessage());
+        skippedTests.put(createDescription(test), test);
+
+    }
+} // XMLJUnitResultFormatter, no newline on purpose
```
Comment (nmancus1) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammar/antlr4/InputAntlr4AstRegressionCassandraInputWithComments.java:411]: Edge case here, namely the file had a single line comment at the end of the file, with no newline. Corresponds to change at https://github.com/checkstyle-antlr4/checkstyle-antlr4/pull/17/commits/2d6255c59e7d82b0945ba1bc47e4eb4e9a1204e8#r661419415

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4 [hunk 9]
```diff
@@ -119,19 +120,38 @@ tokens {
...
+     */
+    public void setCommentListener(CommentListener aCommentListener){
+            mCommentListener = aCommentListener;
+    }
+
+    /** Tracks the starting line of a block comment. */
+    int startLine = -1;
+
+    /** Tracks the starting column of a block comment. */
+    int startCol = -1;
```
Comment (nmancus1) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4:154]: ANTLR4 does not allow local variables in lexer rules now, so we had to make these global.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4 [hunk 10]
```diff
@@ -29,6 +29,7 @@
 
 lexer grammar JavaLexer;
 
+channels { COMMENTS }
```
Comment (nmancus1) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4:32]: This makes ANTLR create a constant for this channel so that we can access it elsewhere as a final static int field , i.e. `JavaLexer.COMMENTS`  (like `JavaAstVisitor).

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java [hunk 11]
```diff
@@ -507,12 +514,38 @@ public void removeChildren() {
...
+        return returnList;
+    }
+
+    /**
+     * Sets the hiddenBefore token field.
+     *
+     * @param hiddenBefore comment token preceding this DetailAstImpl
+     */
+    public void setHiddenBefore(List<Token> hiddenBefore) {
+        this.hiddenBefore = Collections.unmodifiableList(hiddenBefore);
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:540]: No reason for `unmodifiableList` here. You already know that the list is not modified within this class.
Moreover, this is confusing, since the caller actually __can__ modify this unmodifiable list:
```java
import java.util.*;

public class TestClass {
    
    private List<String> hiddenBefore;
    
    public static void main(String... args) {
        var list = new ArrayList();
        list.add("item");
        var testClass = new TestClass();
        testClass.setHiddenBefore(list);
        System.out.println(testClass.hiddenBefore);

        list.clear(); // testClass.hiddenBefore will be "modified" here
        System.out.println(testClass.hiddenBefore);
    }

    public void setHiddenBefore(List<String> hiddenBefore) {
        this.hiddenBefore = Collections.unmodifiableList(hiddenBefore);
    }

}
```

If it is important to have an immutable list here please use `ImmutableList.copyOf`.

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java [hunk 12]
```diff
@@ -507,12 +514,38 @@ public void removeChildren() {
...
+    }
+
+    /**
+     * Sets the hiddenAfter token field.
+     *
+     * @param hiddenAfter comment token following this DetailAstImpl
+     */
+    public void setHiddenAfter(List<Token> hiddenAfter) {
+        this.hiddenAfter = Collections.unmodifiableList(hiddenAfter);
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:549]: same here

File: src/main/java/com/puppycrawl/tools/checkstyle/JavaParser.java [hunk 13]
```diff
@@ -203,16 +209,15 @@ private static DetailAST createSlCommentNode(CommonToken token) {
...
 
-        // column counting begins from 0
-        slComment.setColumnNo(token.getCharPositionInLine() - 1);
+        slComment.setColumnNo(token.getCharPositionInLine());
         slComment.setLineNo(token.getLine());
 
         final DetailAstImpl slCommentContent = new DetailAstImpl();
         slCommentContent.setType(TokenTypes.COMMENT_CONTENT);
 
         // column counting begins from 0
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/JavaParser.java:218]: The comment `column counting begins from 0` is for outdated `- 1` part.  Please remove this line.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FileContents.java [hunk 14]
```diff
@@ -228,8 +228,8 @@ public TextBlock getJavadocBefore(int lineNoBefore) {
         while (lineNo > 0 && (lineIsBlank(lineNo) || lineIsComment(lineNo))) {
             lineNo--;
         }
-
         return javadocComments.get(lineNo);
+
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/FileContents.java:232]: accidentally blank line. plese undo.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4 [hunk 15]
```diff
@@ -119,19 +120,38 @@ tokens {
...
+
+    private CommentListener mCommentListener = null;
+
+    /**
+     * Sets the comment listener so that the CommentListener can
+     * be notified about block and single line comment positions.
+     *
+     */
+    public void setCommentListener(CommentListener aCommentListener){
+            mCommentListener = aCommentListener;
```
Comment (pbludov) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4:147]: Why `m` and `a` prefixes?

**Review by pbludov — COMMENTED** [review]
Some notes

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java (hunk 11 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java:540]: I originally just used `this.hiddenBefore = hiddenBefore;`, but IDEA inspections red-lined it because `Assignment or return of field with mutable type`. This usually results in TC violations. I cannot run `.ci/idea-inspections.sh` locally to verify this, but I will remove it for now. If it is a violation in TC for base repo merge, we will deal with it then.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4 (hunk 15 above)
Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4]: This was just how they were named originally, but there's no need to keep it. Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FileContents.java (hunk 14 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/FileContents.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/JavaParser.java (hunk 13 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/JavaParser.java]: Removed.

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java (hunk 12 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java (hunk 11 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java]: Done.

**Review by pbludov — APPROVED** [review]

**Comment by rnveach** [comment]
@pbludov Don't forget to set next approver.

**Comment by rnveach** [comment]
@nmancus1 I was looking at the AST examples.
https://github.com/checkstyle-antlr4/checkstyle-antlr4/pull/17/commits/28158238cf9053e941a8d231b5713839a7b778f0#diff-d22eb2a454e5ed65fd10f179aa6496334fac0ca40dea36bc4358a13c57108f7cR22-R24
So with ANTLR4, is it not possible to put the comment in a more "correct" position? I find we put the comments in weird locations.

**Comment by nmancus1** [comment]
For context, this is the corresponding code for the AST that you shared:

```java

    /**
     * Javadoc here.
     */
    private static void fooSwitch() {
        ...

```

In the above code, "hiddenBefore" in the context of `private` would consist of `/**\n *\n     * Javadoc here.\n      */` .

All versions of ANTLR handle hidden/ other channels similarly: 

image


In the above diagram, "hiddenBefore" from Channel 1 in the context of `void` would consist of `\n`.

As you can see, any hidden channel tokens after a default channel token, preceding the default channel token in question are considered by ANTLR to be "hiddenBefore".

If I understand you correctly, you mean that "correct" position would be to place the comment on the `METHOD_DEF` node itself. This does make sense, and is totally possible. In the corresponding visit method, we could check an imaginary node's immediate descendants that are tokens (not non-terminal rules) for hidden channel tokens, then add them all to the method definition. We would now need two different visitors, or a bunch of conditional code in the existing visitor. But, we do not differentiate between Javadoc and block comments, so where do we draw the line? While I would say that the above comment is on the method definition, what about this one:

```java

    private static /* block comment here */ void fooSwitch() {
        ...

```

I would say this comment is "on" `void`.

But what about this one:

```java

     /* block comment here */ private static void fooSwitch() {
        ...

```
Is this comment "on" private, or "on'" the method definition?

So, then we need to differentiate between comment types (the best way would be to have a separate channel for each comment type), and decide where they belong based on the type of comment and the context we are in. 

If this is something that we want to do, we would probably want to have two different `JavaAstVisitors`, and just build the AST with or without comments instead of walking the tree and just attaching them to the tokens they are "on" (as we do now). I would say that this could be something we look at after reaching compatibility and merging into the main repo.

**Comment by nmancus1** [comment]
Closing this PR, PR to base repo is opened at https://github.com/checkstyle/checkstyle/pull/10280.
